src/c++/perf_analyzer/client_backend/openai/openai_client.cc

```diff
@@ -0,0 +1,297 @@
+#include <atomic>
+#include <cctype>
+#include <climits>
+#include <cstdint>
+#include <deque>
+#include <iostream>
+#include <string>
+#include <utility>
+
+#include "common.h"
+
+#ifdef TRITON_ENABLE_ZLIB
+#include <zlib.h>
+#endif
+
+extern "C" {
+#include "cencode.h"
+}
+
+#ifdef _WIN32
+#define strncasecmp(x, y, z) _strnicmp(x, y, z)
+#undef min  // NOMINMAX did not resolve std::min compile error
+#endif      //_WIN32
+
+namespace triton { namespace perfanalyzer { namespace clientbackend {
+namespace openai {
+
+//==============================================================================
+
+void
+ChatCompletionRequest::SendResponse(bool is_final, bool is_null)
+{
+  response_callback_(new ChatCompletionResult(
+      http_code_, std::move(response_buffer_), is_final, is_null, request_id_));
+}
+
+ChatCompletionClient::ChatCompletionClient(
+    const std::string& url, const std::string& endpoint, bool verbose,
+    const HttpSslOptions& ssl_options)
+    : HttpClient(std::string(url + "/" + endpoint), verbose, ssl_options)
+{
+}
+
+size_t
+ChatCompletionClient::RequestProvider(
+    void* contents, size_t size, size_t nmemb, void* userp)
+{
+  auto request = reinterpret_cast<ChatCompletionRequest*>(userp);
+
+  size_t input_bytes = 0;
+  request->GetNextInput(
+      reinterpret_cast<uint8_t*>(contents), size * nmemb, &input_bytes);
+
+  if (input_bytes == 0) {
+    request->timer_.CaptureTimestamp(
+        triton::client::RequestTimers::Kind::SEND_END);
+  }
+
+  return input_bytes;
+}
+
+size_t
+ChatCompletionClient::ResponseHeaderHandler(
+    void* contents, size_t size, size_t nmemb, void* userp)
+{
+  auto request = reinterpret_cast<ChatCompletionRequest*>(userp);
+
+  char* buf = reinterpret_cast<char*>(contents);
+  size_t byte_size = size * nmemb;
+
+  std::string hdr(buf, byte_size);
+  std::transform(hdr.begin(), hdr.end(), hdr.begin(), [](unsigned char c) {
+    return std::tolower(c);
+  });
+  if (hdr.find("content-type") != std::string::npos) {
+    request->is_stream_ = (hdr.find("text/event-stream") != std::string::npos);
+  }
+
+  return byte_size;
+}
+
+size_t
+ChatCompletionClient::ResponseHandler(
+    void* contents, size_t size, size_t nmemb, void* userp)
+{
+  // [WIP] verify if the SSE responses received are complete, or the response
+  // need to be stitched first
+  auto request = reinterpret_cast<ChatCompletionRequest*>(userp);
+  if (request->timer_.Timestamp(
+          triton::client::RequestTimers::Kind::RECV_START) == 0) {
+    request->timer_.CaptureTimestamp(
+        triton::client::RequestTimers::Kind::RECV_START);
+  }
+
+  char* buf = reinterpret_cast<char*>(contents);
+  size_t result_bytes = size * nmemb;
+  request->response_buffer_.append(buf, result_bytes);
+  // Send response now if streaming, otherwise wait until request has been
+  // completed
+  if (request->is_stream_) {
+    auto done_signal =
+        (request->response_buffer_.find("data: [DONE]") != std::string::npos);
```

Comment:
Is it documented somewhere (on OpenAI's API reference?) that they have a response with the format you describe here (containing `data: [DONE]`)?

Reply:
This is defined by the SSE format